File: android/app/src/main/java/com/openavc/panel/discovery/ServerInfo.kt
package com.openavc.panel.discovery

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

@Parcelize
data class ServerInfo(
    val name: String,
    val instanceId: String,
    val host: String,
    val port: Int,
    val version: String,
    val panelUrl: String,
    // Default keeps older code paths (manual entry, prefs restore) on plain HTTP.
    // The mDNS path overrides this when the server advertises scheme=https.
    val scheme: String = "http",
) : Parcelable {
    val statusUrl: String get() = "$scheme://$host:$port/api/status"
    val healthUrl: String get() = "$scheme://$host:$port/api/health"

    companion object {
        fun fromPanelUrl(url: String, name: String? = null): ServerInfo? {
            val trimmed = url.trim().trimEnd('/')
            val regex = Regex("^(https?)://([^:/]+)(?::(\\d+))?(/.*)?$", RegexOption.IGNORE_CASE)
            val match = regex.matchEntire(trimmed) ?: return null
            val scheme = match.groupValues[1].lowercase()
            val host = match.groupValues[2]
            // No port in URL → fall back to OpenAVC's defaults (8080 for http, 8443 for https).
            val port = match.groupValues[3]
                .ifEmpty { if (scheme == "https") "8443" else "8080" }
                .toInt()
            val path = match.groupValues[4].ifEmpty { "/panel" }
            return ServerInfo(
                name = name ?: host,
                instanceId = "",
                host = host,
                port = port,
                version = "",
                panelUrl = "$scheme://$host:$port$path",
                scheme = scheme,
            )
        }
    }
}

File: ios/OpenAVCPanel/Discovery/ServerInfo.swift
import Foundation

/// One reachable OpenAVC server, however it was found.
///
/// Deliberately mirrors `ServerInfo.kt` on Android, field for field, so the two
/// apps agree on what a server is and a change to the server contract shows up
/// in one shape on both platforms.
struct ServerInfo: Equatable, Codable, Identifiable {
    let name: String
    let instanceId: String
    let host: String
    let port: Int
    let version: String
    let panelUrl: String
    /// Plain HTTP unless mDNS advertised `scheme=https` or a TLS handoff
    /// upgraded us. Matches the Android default.
    var scheme: String = "http"

    var id: String { instanceId.isEmpty ? "\(host):\(port)" : instanceId }

    var statusUrl: String { "\(scheme)://\(host):\(port)/api/status" }
    var healthUrl: String { "\(scheme)://\(host):\(port)/api/health" }

    /// Parses a panel or pair URL, as scanned from a QR code or typed by hand.
    ///
    /// Port defaults follow OpenAVC's own: 8080 for http, 8443 for https. A URL
    /// with no path is treated as pointing at `/panel`.
    static func fromPanelUrl(_ url: String, name: String? = nil) -> ServerInfo? {
        var trimmed = url.trimmingCharacters(in: .whitespacesAndNewlines)
        while trimmed.hasSuffix("/") { trimmed.removeLast() }

        let pattern = "^(https?)://([^:/]+)(?::(\\d+))?(/.*)?$"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]),
              let match = regex.firstMatch(
                  in: trimmed,
                  options: [.anchored],
                  range: NSRange(trimmed.startIndex..., in: trimmed)
              ),
              match.range.length == trimmed.utf16.count
        else { return nil }

        func group(_ i: Int) -> String {
            guard let r = Range(match.range(at: i), in: trimmed) else { return "" }
            return String(trimmed[r])
        }

        let scheme = group(1).lowercased()
        let host = group(2)
        let portText = group(3)
        let port = Int(portText.isEmpty ? (scheme == "https" ? "8443" : "8080") : portText)
        guard let port, port > 0, port <= 65535 else { return nil }
        let path = group(4).isEmpty ? "/panel" : group(4)

        return ServerInfo(
            name: name ?? host,
            instanceId: "",
            host: host,
            port: port,
            version: "",
            panelUrl: "\(scheme)://\(host):\(port)\(path)",
            scheme: scheme
        )
    }
}
